### ISSUE TITLE
Consider not reporting MA0075 / MA0076 when IFormattable is implemented via explicit interface implementation

### ISSUE THREAD

**Issue by drieseng** [meziantou/Meziantou.Analyzer#365]
**Context**
Rule identifier: _MA0075_
Target Framework: _.NET 6.0_
Language version: _10.0_

**Background**
In .NET 6.0 (and perhaps earlier), **System.Version** implements **ISpanFormattable**. Since **ISpanFormattable** implement **IFormattable**, **System.Version** also had to implement `string IFormattable.ToString(string? format, IFormatProvider? formatProvider)`.

This was done via an explicit interface implementation:
```c#
string IFormattable.ToString(string? format, IFormatProvider? formatProvider) =>
            ToString();
```

Source is available [here](https://github.com/dotnet/runtime/blob/e98fc2a32fd0ea90092d5f93485e2aebd7ec0402/src/libraries/System.Private.CoreLib/src/System/Version.cs#L191-L192).

I suppose the reason why it was done this way, is because there's nothing culture specific in the `ToString()` implementation of **System.Version**. They probably only wanted to implement the `bool TryFormat (Span<char> destination, out int charsWritten, ReadOnlySpan<char> format, IFormatProvider? provider)` method of **ISpanFormattable**, but had no choice.

You could argue that, if a given class only makes available an `ToString(string format, IFormatProvider formatProvider)` method via an explicit interface implementation, that it does not want users to actually invoke it.

**Describe the bug**
When targeting .NET 6.0 or higher, **MA0075** / **MA0076** is reported for the following code fragment:

```c#
Console.WriteLine($"Running on {System.Environment.Version}");
Console.WriteLine("Running on " + System.Environment.Version);
```

There are two solutions:
* Special-case **System.Version** like you already do for other types.
* Do not report **MA0075** / **MA0076** when the type only implements **IFormattable** via an explicit interface implementation (meaning, it has no public `ToString(string format, IFormatProvider formatProvider)` method).

The latter would avoid reporting **MA0075** / **MA0076** for any type that only implements **IFormattable** via an explicit interface implementation, and not just those that are special-cased in [`DoNotUseImplicitCultureSensitiveToStringAnalyzer. IsFormattableType(Compilation compilation, ITypeSymbol typeSymbol)`](https://github.com/meziantou/Meziantou.Analyzer/blob/b6f1cd8721f87655e4a2ae6945155c826154ffbf/src/Meziantou.Analyzer/Rules/DoNotUseImplicitCultureSensitiveToStringAnalyzer.cs#L154-L193).

For example:
```c#
var address = new Address("xx", 50);
Console.WriteLine($"You live at {address}");
Console.WriteLine("You live at " + address);

private class Address : IFormattable
{
    public Address(string street, int houseNumber)
    {
        Street = street;
        HouseNumber = houseNumber;
    }

    public string Street { get; }

    public int HouseNumber { get; }

    public override string ToString()
    {
        return $"{Street}, {HouseNumber}";
    }

    string IFormattable.ToString(string format, IFormatProvider formatProvider)
    {
        return ToString();
    }
}
```

**Comment by drieseng** [comment]
@meziantou Why is **System.TimeSpan** special-cased in [`DoNotUseImplicitCultureSensitiveToStringAnalyzer. IsFormattableType(Compilation compilation, ITypeSymbol typeSymbol)`](https://github.com/meziantou/Meziantou.Analyzer/blob/b6f1cd8721f87655e4a2ae6945155c826154ffbf/src/Meziantou.Analyzer/Rules/DoNotUseImplicitCultureSensitiveToStringAnalyzer.cs#L154-L193)?

**Comment by meziantou** [comment]
I don't know if the rule should exclude types with explicit implementation. For instance, [FormattableString](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/FormattableString.cs,52d4a4f6c1013607) implements the interface explicitly but is culture-sensitive.

Maybe it should check for implicit `ToString` with an [IFormatProvider](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IFormatProvider.cs,0dbf315d83aad6ca) parameter. Or just consider FormattableString as an exception and handle it correctly.

> Why is System.TimeSpan special-cased

If you look at the implementation, the culture is not relevant. That's why a few types are explicitly excluded. Positive numbers are also excluded for the same reason.

**Comment by meziantou** [comment]
- [FontWeight](https://source.dot.net/#PresentationCore/System/Windows/FontWeight.cs,c99876ddb565fb7d,references) implements the interface explicitly but uses the provided culture.
- [KeySpline](https://source.dot.net/#PresentationCore/System/Windows/Media/Animation/KeySpline.cs,470) implements the interface explicitly but provides `ToString(IFormatProvider)`. There are many types with a similar pattern in the .NET BLC

I'm not sure what to do here to avoid false-positive, which are annoying, but still detect potential errors.

**Comment by drieseng** [comment]
> Why is System.TimeSpan special-cased
>> If you look at the implementation, the culture is not relevant. That's why a few types are explicitly excluded. Positive numbers are also excluded for the same reason.

In case of **TimeSpan**, it is relevant:
```c#
var t = new TimeSpan(38, 30, 15);

Console.WriteLine(t.ToString("G", new CultureInfo("en-US")));
Console.WriteLine(t.ToString("G", new CultureInfo("fr-FR")));
```

Result:
```txt
1:14:30:15.0000000
1:14:30:15,0000000
```

**Comment by drieseng** [comment]
> I'm not sure what to do here to avoid false-positive, which are annoying, but still detect potential errors.

I agree, and - in some cases - it's a mystery as to why they did not make these methods public. Except if the explicit interface implementation is considered "for internal use only".

I think it makes sense to report a diagnostic if you use an instance of such a type in:
* a `String.Format(...)` overload that does not take an **IFormatProvider** argument.
* a plain string interpolation (not used as **FormattableString** or not passed a argument to `String.Create(IFormatProvider, DefaultInterpolatedStringHandler)`).

But I'm in doubt whether it's useful to report a diagnostic in other cases.
For example:
```c#
Console.WriteLine("Running on " + System.Environment.Version);
```

Would you expect someone (or a fixer) to transform this into the following:

```c#
Console.WriteLine("Running on " + ((IFormattable) System.Environment.Version).ToString(CultureInfo.CurrentCulture));
```

**Comment by meziantou** [comment]
> In case of TimeSpan, it is relevant:

The philosophy of rule is to report cases where any implicit call to `obj.ToString()` may differ based on the current culture. The rule check if a type implements `IFormattable` as an indication that the current culture may be used in the `ToString` implementation, but there is no guarantee. That's why there are some explicit exclusions to avoid false-positive.

In the case of `TimeSpan`, both `WriteLine` output the same string:

````c#
CultureInfo.CurrentCulture = new CultureInfo("en-US");
Console.WriteLine(t);

CultureInfo.CurrentCulture = new CultureInfo("fr-FR");
Console.WriteLine(t);
````

`System.Version` could be excluded for the same reasons. All components are positive numbers (validated in the ctor), so `ToString()` is culture insensitive.

**Comment by drieseng** [comment]
@meziantou OK, thanks for the clarification. Perhaps consider documenting the list of excludes and "the philosophy of the rule"?
Let me know if you want me to prepare a PR for this (but it would be close to a copy & paste of part of your comment above).

I could add **System.Version** as an exclude in that same PR.

**Comment by meziantou** [comment]
I already made the change for `System.Version` https://github.com/meziantou/Meziantou.Analyzer/commit/58b1ee11d1c69c228e55c291336012eb74bdf25e

Feel free to submit a PR for the doc. If you think other system types should be excluded, you can also submit additional PRs
